Support graceful shutdown on both HTTP/1.1 and HTTP/2

maskit opened the pull request:
For HTTP/1.1, add `Connection: close` header
For HTTP/2, send `GOAWAY` frames

- Generalize the flag name
- Close listening sockets when graceful shutdown has been scheduled
- Send the GOAWAY frame if a response header has`Connection: close`

The last change also fixes an inconsistent behavior that H2 sessions won't be closed even if `Connection: close` header is set by a plugin.


(This still depends on `proxy.config.stop.shutdown_timeout`. I'm going to add support for `traffic_ctl stop --drain` later.)

maskit commented:
@mlibbey This will enable to shed H2 traffic with header_rewrite. (#1693)

maskit commented:
@zizhong Please take a look.

zwoop commented:
@maskit This is a 7.1.0 candidate?

maskit commented:
@zwoop I marked this 8.0 because of necessity but that should be possible. We'll need #1710, #1937 and #2094 if we backport this.

zwoop commented:
@maskit  I trust your judgement here, let me know what you think is the right thing to do. My goal right now is to make 7.1.0 rock solid, and since we're working on other issues, there's still time.

maskit commented:
@zwoop Then I'd say yes.

Although it's not merged yet, we'll need #2096 too.

jpeach commented on proxy/http/HttpProxyServerMain.cc:
```diff
@@ -321,3 +321,10 @@ start_HttpProxyServerBackDoor(int port, int accept_threads)
+{
+  sslNetProcessor.stop_accept();
+  netProcessor.stop_accept();
+}
```
This doesn't work correctly. If you are running under `traffic_manager` then what happens is that the client connection just hangs because `traffic_manager` is still listening on the sockets.

Refusing new connections while draining should absolutely be configurable. Many configurations need to continue to serve clients while the GSLB directs traffic elsewhere. @zwoop.

I saw an assert at one point:
```[jpeach@sk1 trafficserver]$ sudo /opt/ats/bin/traffic_server -T server
traffic_server: using root directory '/opt/ats'
[Jun 10 20:51:40.206] Server {0x7fd9da1b8700} DEBUG: <Main.cc:417 (periodic)> (server) limiting connections based on memory usage has been disabled
[Jun 10 20:52:28.020] Server {0x7fd9da3bc700} DEBUG: <Main.cc:285 (periodic)> (server) received exit signal, shutting down in 30secs
Fatal: UnixNetVConnection.cc:1401: failed assertion `con.fd == NO_FD`
traffic_server: received signal 6 (Aborted)
traffic_server - STACK TRACE:
/opt/ats/bin/traffic_server(_Z19crash_logger_invokeiP9siginfo_tPv+0xc3)[0x504c72]
/lib64/libpthread.so.0(+0x115c0)[0x7fd9ddb065c0]
/lib64/libc.so.6(gsignal+0x9f)[0x7fd9dccb891f]
/lib64/libc.so.6(abort+0x16a)[0x7fd9dccba51a]
/opt/ats/lib/libtsutil.so.8(_Z11ink_warningPKcz+0x0)[0x7fd9df2d58f2]
/opt/ats/lib/libtsutil.so.8(_Z17ats_base64_encodePKhmPcmPm+0x0)[0x7fd9df2d2d65]
/opt/ats/bin/traffic_server(_ZN18UnixNetVConnection4freeEP7EThread+0x2d9)[0x7a8425]
/opt/ats/bin/traffic_server(_ZN18UnixNetVConnection11acceptEventEiP5Event+0x16d)[0x7a7049]
/opt/ats/bin/traffic_server(_ZN12Continuation11handleEventEiPv+0x72)[0x507de6]
/opt/ats/bin/traffic_server(_ZN7EThread13process_eventEP5Eventi+0x134)[0x7ca102]
/opt/ats/bin/traffic_server(_ZN7EThread7executeEv+0x10f)[0x7ca391]
```

maskit replied:
> Refusing new connections while draining should absolutely be configurable. Many configurations need to continue to serve clients while the GSLB directs traffic elsewhere.

I'm fine with keeping sockets open but I think it should have been done before starting graceful shutdown. If we don't close sockets, number of active connections won't goes down without external traffic control.

maskit replied:
> This doesn't work correctly. If you are running under traffic_manager then what happens is that the client connection just hangs because traffic_manager is still listening on the sockets.

Right. Is it OK if I close sockets only if traffic_manager isn't used ? This PR doesn't include support for traffic_manager.

> (This still depends on proxy.config.stop.shutdown_timeout. I'm going to add support for traffic_ctl stop --drain later.)

It was not clear. I mean another Pull Request.

maskit replied:
As for the assert, it seems the NetVConnection has to be closed with `close_UnixNetVConnection()` before calling its free. Otherwise, `net_connections_currently_open_stat` won't be decremented.

jpeach replied:
> If we don't close sockets, number of active connections won't goes down
> without external traffic control.

@maskit That's right. If you are running a CDN you have external load balancing, so you don't need to give errors to clients. I can see the utility of refusing connections, but it should not be the default.

maskit replied:
> If you are running a CDN you have external load balancing
What if you aren't?

jpeach commented on proxy/ProxyClientSession.h:
```diff
@@ -119,6 +121,12 @@ class ProxyClientSession : public VConnection
+  is_draining() const
+  {
+    return http_client_session_draining;
+  }
```
I don't think there's any value in disguising the global variable here. Just use it directly where you need it.

maskit replied:
I wrapped it because the variable will be replaced with a metric you suggested.

jpeach commented on proxy/ProxyClientSession.h:
```diff
@@ -30,6 +30,8 @@
 #include "InkAPIInternal.h"
 #include "http/HttpServerSession.h"
 
+extern volatile bool http_client_session_draining;
```
There should be a metric that publishes this state so that operators can tell we are in draining state.

maskit replied:
Can I left that as a future work? Because main purpose of this PR is generalizing the flag and supporting graceful shutdown on HTTP/1.1. This is just a replacement of `http2_drain`.

jpeach commented on proxy/http2/Http2ConnectionState.h:
```diff
@@ -38,7 +38,7 @@ enum Http2SendADataFrameResult {
 };
 
-enum Http2ShutdownState { NOT_INITIATED, INITIATED, IN_PROGRESS };
+enum Http2ShutdownState { NOT_PLANNED, NOT_INITIATED, INITIATED, IN_PROGRESS };
```
What do these states mean? I can't figure out from the usage what you intend here.

jpeach replied:
These enums should be renamed to be consistent with the rest of the code. I think that all you need are:

```C
enum Http2ShutdownState {
    HTTP2_SHUTDOWN_NONE,
    HTTP2_SHUTDOWN_BEGIN,
    HTTP2_SHUTDOWN_INPROGRESS
};
```

jpeach replied:
Looking at the code for handling `HTTP2_SESSION_EVENT_SHUTDOWN_INIT`, is it safe to schedule the `HTTP2_SESSION_EVENT_SHUTDOWN_CONT` in 2 seconds? What guarantees that the `Http2ConnectionState` is still alive at that time?

maskit replied:
NOT_PLANNED: Shutdown is not planned (it'll keep working)
NOT_INITIATED: Shutdown is planned but nothing is happened yet (The 1st GOAWAY will be sent)
INITIATED: The first GOAWAY is sent and waiting the 2sec before sending the 2nd GOAWAY
IN_PROGRESS: The second GOAWAY is sent and waiting shutdown timeout

I'll rename them. Let me know if you have better names.

maskit replied:
Not sure, but you can say the same thing for other events, it should be guaranteed though. It's not special.

jpeach replied:
Let's start by making them consistent with other enums, so:
```C
HTTP2_SHUTDOWN_NONE,
HTTP2_SHUTDOWN_NOT_INITIATED,
HTTP2_SHUTDOWN_INITIATED,
HTTP2_SHUTDOWN_IN_PROGRESS
```

Would you be able to make another PR later that reduces this down to just 3 values?

jpeach replied:
There's only one other `schedule_in` in HTTP/2 and that keeps hold of the returned `Action` so it can be canceled. I don't this this usage is safe.

maskit replied:
It should be possible if I change the conditions, but I guess the condition will be a bit complex.
I'll make them consistent for now.

maskit replied:
I don't get it. Are you suggesting any of these?
- Keep `Http2ConnectionState` alive until the event processed
- Make it cancelable and cancel the event when `Http2ConnectoinState` dies

Is it guaranteed that `Http2ConnectionState` is still alive when  `HTTP2_SESSION_EVENT_RECV` is going to be processed?

jpeach replied:
I just don't understand how this `schedule_in` is safe. If the object is destroyed before the 2sec timeout, the event will run on a freed object AFAICT. I think you have to make it cancelable and cancel the event when Http2ConnectionState dies.

The `HTTP2_SESSION_EVENT_RECV` is different, since `send_connection_event()` delivers the event directly, so it is reentrant on the event handler.

maskit replied:
Ok, I understand your point.

`HTTP2_SESSION_EVENT_RECV ` was a bad example :) I wanted to know how other places guarantee the continuation is still alive. Anyway, I'll look into it and try to make it solid.

jpeach commented on proxy/http2/Http2Stream.h:
```diff
@@ -146,6 +146,12 @@ class Http2Stream : public ProxyClientTransaction
   void reenable(VIO *vio);
   virtual void transaction_done();
+  virtual bool
+  ignore_keep_alive()
```
```C
bool ignore_keep_alive() override { ... }
```

maskit replied:
Will do after #2113. (Many warnings show up.)

jpeach replied:
Thanks!

jpeach commented on proxy/http2/Http2Stream.h:
```diff
@@ -146,6 +146,12 @@ class Http2Stream : public ProxyClientTransaction
+  virtual bool
+  ignore_keep_alive()
+  {
+    return false;
```
This needs a comment explaining it.

maskit replied:
Honestly, I don't understand why it was true. Without this change, `Connection: close` is always added (and is removed while converting headers from 1.1 to 2). I think it should be handled as the same as HTTP/1.1.

jpeach replied:
OK, sounds good. Just comment the explanation?

maskit replied:
Got it.

jpeach commented:
@zwoop @maskit I don't think this is a backport candidate until the whole feature is complete. I don't want us to get stuck on something just for compatibility reasons.

maskit commented:
Even if I couldn't complete the whole feature, I think most changes we have already merged on master (and code on this PR) is still valid. HTTP/2 sessions should be closed if (proxy's) response headers have `Connection: close`. I'd rather remove only the shutdown part than postpone all the changes to 8.0.

maskit commented:
Done
* Made enum names consistent
* Added `override`
* Explained why `ignore_keep_alive()` should return `false`
* Removed code that closes listening sockets
  * It refuses new connections on SessionAccept instead (and it's not configurable for now)

WIP
* Cancel HTTP2_SESSION_EVENT_SHUTDOWN_CONT event when Http2ConnectionState dies 

Need discussion
* Whether refuse new connections while draining

maskit commented:
I don't see any reason to not refuse new connections while draining. Refusing new connections might be overkill for users who use load balancer, but if ATS doesn't refuse connections, draining feature will be only for users who use load balancer (or some external controller). You can remove servers from load balancer before starting draining so I think it shouldn't be a problem even if the behavior is not configurable.

zwoop commented:
Moving out to 8.0.0 for now, since there seems to be some contention on this topic. :)

jpeach commented:
> On Jun 13, 2017, at 1:10 AM, Masakazu Kitajo <notifications@github.com> wrote:
> 
> I don't see any reason to not refuse new connections while draining.

It definitely has to be configurable.

> Refusing new connections might be overkill for users who use load balancer, but if ATS doesn't refuse connections, draining feature will be only for users who use load balancer (or some external controller).

That's everyone. You have to manage getting the traffic into ATS somehow :) Also, if you refuse connections you have to actually refuse them; you can't leave them to time out in the listen backlog and you can't shut them down once you have accepted them.

> You can remove servers from load balancer before starting draining so I think it shouldn't be a problem even if the behavior is not configurable.

Not if you are using a DNS-based GSLB.

maskit commented:
> You have to manage getting the traffic into ATS somehow :) 

Can we assume that all ATS operators can control it? If the answer was yes, why is it need to be configurable?

> Also, if you refuse connections you have to actually refuse them;

I agree. That's why I tried to close listening sockets. So, is it OK to close the sockets from TS if TS isn't managed by TM (and users want it) ?

> Not if you are using a DNS-based GSLB.

Why not? Isn't it possible to remove servers? I'm not familiar with DNS-based GLSB.

bryancall commented:
[approve ci autest]

jpeach commented:
@maskit When you have a DNS GSLB clients will continue to send traffic to servers as the TTL drains. In any kind of GSLB solution it is reasonable to not want to send an error to the client while traffic is draining. I definitely think that dropping connections must be configurable.

maskit commented:
@jpeach I agree that we should not send errors while draining. However, in any kind of GSLB solutions, I think we can avoid the errors if we remove servers from GSLB before starting drain, can't we?

jpeach commented:
> On Jun 21, 2017, at 3:53 PM, Masakazu Kitajo <notifications@github.com> wrote:
> 
> @jpeach I agree that we should not send errors while draining. However, in any kind of GSLB solutions, I think we can avoid the errors if we remove servers from GSLB before starting drain, can't we?

I just pointed out that this can't happen for DNS-based GSLB. If removing from every kind of GSLB guaranteed that no more clients would connect, then why are you closing connections?

maskit commented:
Even DNS-based GSLB, we can do that after the TTL, no? A reason why I close listening sockets is to enable ATS to perform graceful shutdown standalone (without external controllers).

So, if we *require* external controllers when users want to shutdown ATS gracefully, then I'm fine with keeping the sockets open.

jpeach commented:
> On Jun 22, 2017, at 9:38 AM, Masakazu Kitajo <notifications@github.com> wrote:
> 
> Even DNS-based GSLB, we can do that after the TTL, no?
> 
No, clients will never drain to 0.
> A reason why I close listening sockets is to enable ATS to perform graceful shutdown standalone (without external controllers).
> 
That's fine, it just needs to be configurable.
> So, if we require external controllers when users want to shutdown ATS gracefully, then I'm fine with keeping the sockets open.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

maskit commented:
> No, clients will never drain to 0.

OK, fair enough. Thank you for your input. I'll make it configurable.

maskit commented:
Updated the code.

- Generalize the flag name
- Add "Connection: close" automatically while draining traffic (for H1)
- Send the GOAWAY frame if a response header has "Connection: close" (for H2)
- Enable to close listening sockets while draining traffic by a setting "proxy.config.restart.stop_listening" (default behavior is keeping them open)

Support for "traffic_ctl stop --drain" will be another PR. I'll send it within a few days with WIP label so that we can agree on the design and the goal.

maskit commented:
Next step will be something like this: https://github.com/maskit/trafficserver/pull/1.

bryancall commented:
[approve ci autest]

zwoop commented:
@maskit @masaori335 Where are we with this? Is this still a WIP, or getting ready for review / merge? If so, please remove the WIP label.

maskit commented:
It's ready for review.

bryancall commented:
This needs to be reviewed.

zizhong commented:
I'm thinking about adding tests. @maskit do you have any good idea about how to test this?

maskit commented:
@zizhong I think autest would be an option.

We would need an origin server ( or a plugin) which returns a response with few seconds of delay without `Connection: close`.

The test would be something like:
1. Set `proxy.config.stop.shutdown_timeout` to 10 sec
1. Launch Traffic Server
1. Send an HTTP request
1. Send SIGTERM to the process
1. Check if the response has `Connection: close`
1. Wait for 10 sec
1. Check if the process doesn't exist

I think we don't have any ways to test this behavior on HTTP/2.

zizhong commented:
@maskit Something more useful is that ATS sends the GOAWAY frame if a response header has `Connection: close` no matter the draining started or not. This will enable a plugin to initiate the graceful shutdown process for both HTTP/1 and HTTP/2.

maskit commented:
@zizhong That logic is already implemented :)
https://github.com/apache/trafficserver/pull/2106/files#diff-83845675d9d6c1da389e9e0cdbe4ceb3R565

maskit commented:
Rebased on the latest master. The code hasn't been changed since Jul 20. I just resolved conflicts.

maskit commented:
Made a small change.
```diff
       RecInt timeout = 0;
-      if (RecGetRecordInt("proxy.config.stop.shutdown_timeout", &timeout) == REC_ERR_OKAY && timeout) {
+      if (RecGetRecordInt("proxy.config.stop.shutdown_timeout", &timeout) == REC_ERR_OKAY && timeout &&
+          !http_client_session_draining) {
         http_client_session_draining = true;
```

maskit commented:
@jpeach Can you take another look?

maskit commented:
[approve ci]

zizhong commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -411,6 +411,19 @@ Thread Variables
    This setting specifies the number of active client connections
    for use by :option:`traffic_ctl server restart --drain`.
 
+.. ts:cv:: CONFIG proxy.config.restart.stop_listening INT 0
```
@maskit 
Tests are needed for both HTTP/1 draining and HTTP/2 graceful shutdown. They can help you when you add "traffic_ctl --drain" feature.
For how to test it in HTTP/2, you can check out this gist. https://gist.github.com/zizhong/401bf0618c0a977d0cd7e1042ee632fc

maskit replied:
Thanks for the test. Nice job. I tested it manually but we should automate it, and the test is helpful even without supporting "traffic_ctl --drain". Can you make the test autest compatible and submit a PR?

Also, as you know, the GOAWAY frame for graceful shutdown has the unique stream id. So, it would be great if the test checks the stream id.

zizhong replied:
Sure. Will do.

maskit commented:
@jpeach I dismissed your change request because I think they are already addressed. Please file issues if there's something I missed.
